# The Effects of Astronomical Events on Bitcoin Prices

#### Hatice Kübra Tekin 21946611
#### Feyza Özkan 2200356815



In this project,we are investigating whether various astronomical events such as eclipses ,meteor showers, equinoxes affects bitcoin prices with the help of statistical testing methods.


## Table of Content

[Problem](#problem)   
[Data Collection](#data_collection)     
[Data Preparation](#data_preparation)   
[Exploratory Data Analysis (EDA)](#EDA)   
[Feature Engineering](#fe)   
[Analyzing ](#anl)   
[Conclusion](#conc)   
[References](#references)   




## Problem <a class="anchor" id="problem"></a>

It is known that bitcoin prices are too volatile, they can be affected by many factors;the supply and demand of BTC, competition from other cryptocurrencies and news,political issues, cost of production and regulation etc.Also,in the Internet ,there are many articles that claims eclipses affect bitcoin prices;however,we did not encountered any detailed investigation about that claim despite it is an hot-topic in the cryptocurrency area.

   In this Data Science Capstone Project,we analyze whether this claim is true or not. In addition to eclipses, we also decide to examine statistically meteor showers and equinoxes. In this way, we will see that if stated astronomical events affect bitcoin prices or not.

## Data Collection<a class="anchor" id="data_collection"></a>

## In project,we have used two main data sets.
1. BTC HISTORY  
We obtain BTC data using yfinance API(open source tool that can be used for educational purposes).Data consists of the dates between 2014-09-17 and 2022-12-31,and open,high,low,close prices in USD,finally  Volume	Dividends	Stock Splits

2. EVENTS HISTORY  
We have created events.csv files by gathering dates information about  lunar eclipses,solar eclipses ,meteor showers, and equinoxes using online sources.First two columns of data set shows the starting and ending dates of event, third column contains event names, fourth column contains initials of events,and the last two column shows events that initials represent.


In [1]:
#downloading necessary finance packages
!pip install yfinance

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 50 kB 2.8 MB/s 
     |████████████████████████████████| 128 kB 7.6 MB/s 
     |████████████████████████████████| 4.0 MB 61.4 MB/s 
     |████████████████████████████████| 110 kB 59.3 MB/s 
     |████████████████████████████████| 62 kB 1.5 MB/s 
     |████████████████████████████████| 112 kB 51.0 MB/s 
  Attempting uninstall: requests
    Found existing installation: requests 2.23.0
    Uninstalling requests-2.23.0:
      Successfully uninstalled requests-2.23.0
  Attempting uninstall: html5lib
    Found existing installation: html5lib 1.0.1
    Uninstalling html5lib-1.0.1:
      Successfully uninstalled html5lib-1.0.1
  Attempting uninstall: beautifulsoup4
    Found existing installation: beautifulsoup4 4.6.3
    Uninstalling beautifulsoup4-4.6.3:
      Successfully uninstalled beautifulsoup4-4.6.3


In [2]:
#importing libraries
import pandas as pd
import yfinance as yf

In [3]:
#initializing time interval for max historical data
btc = yf.Ticker("BTC-USD")
df = btc.history(period="max")

In [4]:
#taking historical data until the end of 2022
df = df.iloc[:3028,:]

In [5]:
#reading events csv file
events = pd.read_csv("events.csv")

In [6]:
#top 5 row of our df dataframe
df.head()

Open        High         Low       Close  \
Date                                                                        
2014-09-17 00:00:00+00:00  465.864014  468.174011  452.421997  457.334015   
2014-09-18 00:00:00+00:00  456.859985  456.859985  413.104004  424.440002   
2014-09-19 00:00:00+00:00  424.102997  427.834991  384.532013  394.795990   
2014-09-20 00:00:00+00:00  394.673004  423.295990  389.882996  408.903992   
2014-09-21 00:00:00+00:00  408.084991  412.425995  393.181000  398.821014   

                             Volume  Dividends  Stock Splits  
Date                                                          
2014-09-17 00:00:00+00:00  21056800        0.0           0.0  
2014-09-18 00:00:00+00:00  34483200        0.0           0.0  
2014-09-19 00:00:00+00:00  37919700        0.0           0.0  
2014-09-20 00:00:00+00:00  36863600        0.0           0.0  
2014-09-21 00:00:00+00:00  26580100        0.0           0.0

In [7]:
#top 5 rows of our events dataframe
events.head()

START DATE   END DATE                      EVENT TYPE  Unnamed: 4  \
0   1/2/2014   1/3/2014  Quadrantids Meteor Shower   MS         NaN   
1  3/20/2014  3/20/2014              March Equinox    E         NaN   
2  4/15/2014  4/15/2014        Total Lunar Eclipse   LE         NaN   
3  4/22/2014  4/23/2014       Lyrids Meteor Shower   MS         NaN   
4  4/29/2014  4/29/2014      Annular Solar Eclipse   SE         NaN   

   Unnamed: 5  Unnamed: 6     Unnamed: 7 Unnamed: 8  
0         NaN         NaN  Meteor Shower         MS  
1         NaN         NaN        Equinox          E  
2         NaN         NaN  Lunar Eclipse         LE  
3         NaN         NaN  Solar Eclipse         SE  
4         NaN         NaN            NaN        NaN

In [8]:
#getting a concise summary of the df dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 3028 entries, 2014-09-17 00:00:00+00:00 to 2022-12-31 00:00:00+00:00
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Open          3028 non-null   float64
 1   High          3028 non-null   float64
 2   Low           3028 non-null   float64
 3   Close         3028 non-null   float64
 4   Volume        3028 non-null   int64  
 5   Dividends     3028 non-null   float64
 6   Stock Splits  3028 non-null   float64
dtypes: float64(6), int64(1)
memory usage: 189.2 KB


In [9]:
#generating descriptive statistics
df.describe(include= "all")

Open          High           Low         Close        Volume  \
count   3028.000000   3028.000000   3028.000000   3028.000000  3.028000e+03   
mean   12858.371721  13182.565605  12495.418786  12862.178276  1.634383e+10   
std    16190.317021  16604.605313  15710.663934  16185.351139  1.997501e+10   
min      176.897003    211.731003    171.509995    178.102997  5.914570e+06   
25%      672.789017    684.548492    663.220993    675.556259  1.006505e+08   
50%     7097.275635   7287.041260   6894.411133   7106.541992  7.367460e+09   
75%    16804.387207  17089.460938  16530.455566  16803.772949  2.792719e+10   
max    67549.734375  68789.625000  66382.062500  67566.828125  3.509679e+11   

       Dividends  Stock Splits  
count     3028.0        3028.0  
mean         0.0           0.0  
std          0.0           0.0  
min          0.0           0.0  
25%          0.0           0.0  
50%          0.0           0.0  
75%          0.0           0.0  
max          0.0           0.0

In [10]:
#getting a concise summary of the events dataframe
events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 0 to 157
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   START DATE  158 non-null    object 
 1   END DATE    158 non-null    object 
 2   EVENT       158 non-null    object 
 3   TYPE        158 non-null    object 
 4   Unnamed: 4  0 non-null      float64
 5   Unnamed: 5  0 non-null      float64
 6   Unnamed: 6  0 non-null      float64
 7   Unnamed: 7  4 non-null      object 
 8   Unnamed: 8  4 non-null      object 
dtypes: float64(3), object(6)
memory usage: 11.2+ KB


## Data Preparation<a class="anchor" id="data_preparation"></a>

In this part we transformed our data frames into correct format that we want

In [11]:
#since date is seen as index,move it into new column 
df.reset_index(inplace=True)
df

Date          Open          High           Low  \
0    2014-09-17 00:00:00+00:00    465.864014    468.174011    452.421997   
1    2014-09-18 00:00:00+00:00    456.859985    456.859985    413.104004   
2    2014-09-19 00:00:00+00:00    424.102997    427.834991    384.532013   
3    2014-09-20 00:00:00+00:00    394.673004    423.295990    389.882996   
4    2014-09-21 00:00:00+00:00    408.084991    412.425995    393.181000   
...                        ...           ...           ...           ...   
3023 2022-12-27 00:00:00+00:00  16919.291016  16959.845703  16642.072266   
3024 2022-12-28 00:00:00+00:00  16716.400391  16768.169922  16497.556641   
3025 2022-12-29 00:00:00+00:00  16552.322266  16651.755859  16508.683594   
3026 2022-12-30 00:00:00+00:00  16641.330078  16643.427734  16408.474609   
3027 2022-12-31 00:00:00+00:00  16608.707031  16622.609375  16517.750000   

             Close       Volume  Dividends  Stock Splits  
0       457.334015     21056800        0.0           0.0  
1       424.440002     34483200        0.0           0.0  
2       394.795990     37919700        0.0           0.0  
3       408.903992     36863600        0.0           0.0  
4       398.821014     26580100        0.0           0.0  
...            ...          ...        ...           ...  
3023  16717.173828  15748580239        0.0           0.0  
3024  16552.572266  17005713920        0.0           0.0  
3025  16642.341797  14472237479        0.0           0.0  
3026  16602.585938  15929162910        0.0           0.0  
3027  16547.046875  11240764416        0.0           0.0  

[3028 rows x 8 columns]

In [12]:
#checking for null values 
df.isnull().values.any()

False

In [13]:
#checking for duplicated values
df.duplicated().values.any()

False

In [14]:
#displaying data types
df.dtypes

Date            datetime64[ns, UTC]
Open                        float64
High                        float64
Low                         float64
Close                       float64
Volume                        int64
Dividends                   float64
Stock Splits                float64
dtype: object

In [15]:
#reformatting Date column from [ns, UTC] to [ns] for simplicity
df['Date'] = df['Date'].dt.tz_localize(None)

In [16]:
df.dtypes

Date            datetime64[ns]
Open                   float64
High                   float64
Low                    float64
Close                  float64
Volume                   int64
Dividends              float64
Stock Splits           float64
dtype: object

In [17]:
#checking for null values 
events.isnull().values.any()

True

In [18]:
#checking for duplicated values
events.duplicated().values.any()

False

In [19]:
#dropping columns that contains missing values
events.drop(["Unnamed: 4","Unnamed: 5","Unnamed: 6","Unnamed: 7","Unnamed: 8"],axis=1,inplace=True)
events

START DATE    END DATE                      EVENT TYPE
0      1/2/2014    1/3/2014  Quadrantids Meteor Shower   MS
1     3/20/2014   3/20/2014              March Equinox    E
2     4/15/2014   4/15/2014        Total Lunar Eclipse   LE
3     4/22/2014   4/23/2014       Lyrids Meteor Shower   MS
4     4/29/2014   4/29/2014      Annular Solar Eclipse   SE
..          ...         ...                        ...  ...
153   11/4/2022   11/5/2022     Traurids Meteor Shower   MS
154   11/8/2022   11/8/2022        Total Lunar Eclipse   LE
155  11/17/2022  11/18/2022      Leonids Meteor Shower   MS
156  12/13/2022  12/14/2022     Geminids Meteor Shower   MS
157  12/21/2022  12/22/2022       Ursids Meteor Shower   MS

[158 rows x 4 columns]

In [20]:
#again checking for null values
events.isnull().values.any()

False

In [21]:
events.dtypes

START DATE    object
END DATE      object
EVENT         object
TYPE          object
dtype: object

In [22]:
#reformatting dates from object data type to [ns] 
events['START DATE'] = pd.to_datetime(events['START DATE'])
events['END DATE'] = pd.to_datetime(events['END DATE'])

In [23]:
events.dtypes

START DATE    datetime64[ns]
END DATE      datetime64[ns]
EVENT                 object
TYPE                  object
dtype: object

## Exploratory Data Analysis (EDA)<a class="anchor" id="EDA"></a>

In this part we tried to understand data that we deal with.

In [24]:
#printing column names of our dataframes
print("df columns: " ,df.columns)
print("\nevents columns: " ,events.columns)

df columns:  Index(['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Dividends',
       'Stock Splits'],
      dtype='object')

events columns:  Index(['START DATE', 'END DATE', 'EVENT', 'TYPE'], dtype='object')


In [25]:
#final data types of df and events data frame
print(df.dtypes,"\n----------------------------")
print(events.dtypes)


Date            datetime64[ns]
Open                   float64
High                   float64
Low                    float64
Close                  float64
Volume                   int64
Dividends              float64
Stock Splits           float64
dtype: object 
----------------------------
START DATE    datetime64[ns]
END DATE      datetime64[ns]
EVENT                 object
TYPE                  object
dtype: object


In [26]:
#showing all columns are in correct format, and data frame does not contain any null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3028 entries, 0 to 3027
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Date          3028 non-null   datetime64[ns]
 1   Open          3028 non-null   float64       
 2   High          3028 non-null   float64       
 3   Low           3028 non-null   float64       
 4   Close         3028 non-null   float64       
 5   Volume        3028 non-null   int64         
 6   Dividends     3028 non-null   float64       
 7   Stock Splits  3028 non-null   float64       
dtypes: datetime64[ns](1), float64(6), int64(1)
memory usage: 189.4 KB


In [27]:
#generating descriptive statistics
df.describe()

Open          High           Low         Close        Volume  \
count   3028.000000   3028.000000   3028.000000   3028.000000  3.028000e+03   
mean   12858.371721  13182.565605  12495.418786  12862.178276  1.634383e+10   
std    16190.317021  16604.605313  15710.663934  16185.351139  1.997501e+10   
min      176.897003    211.731003    171.509995    178.102997  5.914570e+06   
25%      672.789017    684.548492    663.220993    675.556259  1.006505e+08   
50%     7097.275635   7287.041260   6894.411133   7106.541992  7.367460e+09   
75%    16804.387207  17089.460938  16530.455566  16803.772949  2.792719e+10   
max    67549.734375  68789.625000  66382.062500  67566.828125  3.509679e+11   

       Dividends  Stock Splits  
count     3028.0        3028.0  
mean         0.0           0.0  
std          0.0           0.0  
min          0.0           0.0  
25%          0.0           0.0  
50%          0.0           0.0  
75%          0.0           0.0  
max          0.0           0.0

In [28]:
#showing all columns in correct format,and data frame does not contain any null values
events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 0 to 157
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   START DATE  158 non-null    datetime64[ns]
 1   END DATE    158 non-null    datetime64[ns]
 2   EVENT       158 non-null    object        
 3   TYPE        158 non-null    object        
dtypes: datetime64[ns](2), object(2)
memory usage: 5.1+ KB


In [29]:
#importing necessary library for graph BTC price in USD.
import plotly.graph_objects as go

In [30]:
fig = go.Figure(data=[go.Candlestick(x=df['Date'],
                open=df['Open'],
                high=df['High'],
                low=df['Low'],
                close=df['Close'])],
                
                layout=go.Layout(title='BTC-USDT GRAPH' ,
                yaxis_title='BTC Stock',
                legend_title_text = "Events")
               )



#displaying btc-usdt graph over the years
fig.show()


In [31]:
#showing four astronomical events on the graph
fig.add_trace(go.Scatter(x=events[events["TYPE"] == "MS"].iloc[5:,0],
                         y = df[df['Date'].isin(events["START DATE"].loc[events["TYPE"]== "MS"]) ].iloc[:,1],
                         marker_color= "rgb(249, 5, 208)",
                         name = "Meteor Shower",
                         mode="markers",
                         marker_size = 7))

fig.add_trace(go.Scatter(x=events[events["TYPE"] == "E"].iloc[1:,0],
                         y = df[df['Date'].isin(events["START DATE"].loc[events["TYPE"]== "E"]) ].iloc[:,1],
                         marker_color= "rgb(223, 226, 33)",
                         marker_symbol = 'diamond',
                         name = "Equinox",
                         mode="markers",
                         marker_size = 7))

fig.add_trace(go.Scatter(x=events[events["TYPE"] == "LE"].iloc[1:,0],
                         y = df[df['Date'].isin(events["START DATE"].loc[events["TYPE"]== "LE"]) ].iloc[:,1],
                         marker_color= "rgb(13, 79, 232)",
                         marker_symbol = 'star',
                         name = "Lunar Eclipse",
                         mode="markers",
                         marker_size = 7))

fig.add_trace(go.Scatter(x=events[events["TYPE"] == "SE"].iloc[1:,0],
                         y = df[df['Date'].isin(events["START DATE"].loc[events["TYPE"]== "SE"]) ].iloc[:,1],
                         marker_color= "rgb(0,0,0)",
                         marker_symbol = 'square',
                         name = "Solar Eclipse",
                         mode="markers",
                         marker_size = 7))

fig.show()


## Feature Engineering<a class="anchor" id="fe"></a>


In this part, we ensured that only the necessary columns were left in the data frames. We dropped unnecessary rows from the "events" data frame and created separate data frames for each type in the "events" data frame.

In [32]:
# dropped not necessary columns
price= df.drop(["High","Low","Volume","Dividends","Stock Splits"], axis=1)   
price.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3028 entries, 0 to 3027
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Date    3028 non-null   datetime64[ns]
 1   Open    3028 non-null   float64       
 2   Close   3028 non-null   float64       
dtypes: datetime64[ns](1), float64(2)
memory usage: 71.1 KB


In [33]:
# dropped unnecessary columns and rows
events.drop(["EVENT"],axis=1,inplace=True)
events.drop(index=events.index[:8],inplace=True)
events.reset_index(drop=True,inplace=True)
events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   START DATE  150 non-null    datetime64[ns]
 1   END DATE    150 non-null    datetime64[ns]
 2   TYPE        150 non-null    object        
dtypes: datetime64[ns](2), object(1)
memory usage: 3.6+ KB


In [34]:
# split the "events" data frame according to their type
lunar_eclipse_df = events[events["TYPE"] == 'LE'].reset_index(drop=True)
print("Lunar Eclipse:\n",lunar_eclipse_df.head(),"\n----------------------------")
solar_eclipse_df = events[events["TYPE"] == 'SE'].reset_index(drop=True)
print("Solar Eclipse:\n",solar_eclipse_df.head(),"\n----------------------------")
meteor_shower_df = events[events["TYPE"] == 'MS'].reset_index(drop=True)
print("Meteor Shower:\n",meteor_shower_df.head(),"\n----------------------------")
equinox_df = events[events["TYPE"] == 'E'].reset_index(drop=True)
print("Equinox:\n",equinox_df.head())

Lunar Eclipse:
   START DATE   END DATE TYPE
0 2014-10-08 2014-10-08   LE
1 2015-04-04 2015-04-04   LE
2 2015-09-28 2015-09-28   LE
3 2016-03-09 2016-03-09   LE
4 2016-03-23 2016-03-23   LE 
----------------------------
Solar Eclipse:
   START DATE   END DATE TYPE
0 2014-10-23 2014-10-23   SE
1 2015-03-20 2015-03-20   SE
2 2015-09-13 2015-09-13   SE
3 2016-09-01 2016-09-01   SE
4 2017-02-26 2017-02-26   SE 
----------------------------
Meteor Shower:
   START DATE   END DATE TYPE
0 2014-10-08 2014-10-09   MS
1 2014-10-20 2014-10-21   MS
2 2014-11-05 2014-11-06   MS
3 2014-11-17 2014-11-18   MS
4 2014-12-13 2014-12-14   MS 
----------------------------
Equinox:
   START DATE   END DATE TYPE
0 2014-09-23 2014-09-23    E
1 2015-03-20 2015-03-20    E
2 2015-09-23 2015-09-23    E
3 2016-03-20 2016-03-20    E
4 2016-09-22 2016-09-22    E


## Analyzing<a class="anchor" id="anl"></a>

In this part, we test the data with the help of Pearson's Chi-Square statistical hypothesis after calculating the increase or decrease in bitcoin prices on the dates of astronomical events.

In [35]:
# This function adds two columns to the given data frame; the amount of the BTC price change and increase or decrease.
def final_df(event, btc):
  changes_price = []
  inc_or_dec = []     # 1(increase) and 0(decrease)
  btc.set_index("Date",inplace=True)
  for i in range(len(event)):
    changes_price.append((btc["Close"][event["END DATE"][i]]-btc["Open"][event["START DATE"][i]]))
    if ((btc["Close"][event["END DATE"][i]]-btc["Open"][event["START DATE"][i]])>0):
      inc_or_dec.append(1)
    else:
      inc_or_dec.append(0)
  df1 = pd.DataFrame(list(zip(changes_price,inc_or_dec)), columns =['Changes in Price', 'Inc or Dec'])
  new_df = pd.concat([event["START DATE"],event["END DATE"],df1,event["TYPE"]],axis=1)
  return new_df

In [36]:
# calculated and added changes in price, increase or decrease 
ms_df = final_df(meteor_shower_df,price.copy(deep=True))
ms_df

START DATE   END DATE  Changes in Price  Inc or Dec TYPE
0  2014-10-08 2014-10-09         28.910004           1   MS
1  2014-10-20 2014-10-21         -2.755981           0   MS
2  2014-11-05 2014-11-06         18.606995           1   MS
3  2014-11-17 2014-11-18        -13.151001           0   MS
4  2014-12-13 2014-12-14         -0.749023           0   MS
..        ...        ...               ...         ...  ...
90 2022-10-21 2022-10-22        154.986328           1   MS
91 2022-11-04 2022-11-05       1073.921875           1   MS
92 2022-11-17 2022-11-18         27.351562           1   MS
93 2022-12-13 2022-12-14        609.208984           1   MS
94 2022-12-21 2022-12-22        -74.185547           0   MS

[95 rows x 5 columns]

In [37]:
e_df = final_df(equinox_df ,price.copy(deep=True))
e_df

START DATE   END DATE  Changes in Price  Inc or Dec TYPE
0  2014-09-23 2014-09-23         33.698975           1    E
1  2015-03-20 2015-03-20          0.792999           1    E
2  2015-09-23 2015-09-23         -0.653000           0    E
3  2016-03-20 2016-03-20          3.354004           1    E
4  2016-09-22 2016-09-22         -0.981018           0    E
5  2017-03-20 2017-03-20         16.989990           1    E
6  2017-09-22 2017-09-22          2.679932           1    E
7  2018-03-20 2018-03-20        293.799805           1    E
8  2018-09-23 2018-09-23         -4.689941           0    E
9  2019-03-21 2019-03-21        -54.626953           0    E
10 2019-09-23 2019-09-23       -338.638672           0    E
11 2020-03-20 2020-03-20          7.124512           1    E
12 2020-09-22 2020-09-22         78.835938           1    E
13 2021-03-20 2021-03-20        -18.617188           0    E
14 2021-09-22 2021-09-22       2896.554688           1    E
15 2022-03-20 2022-03-20       -943.582031           0    E
16 2022-09-23 2022-09-23       -114.761719           0    E

In [38]:
se_df = final_df(solar_eclipse_df,price.copy(deep=True))
se_df

START DATE   END DATE  Changes in Price  Inc or Dec TYPE
0  2014-10-23 2014-10-23        -24.545013           0   SE
1  2015-03-20 2015-03-20          0.792999           1   SE
2  2015-09-13 2015-09-13         -4.730011           0   SE
3  2016-09-01 2016-09-01         -3.243042           0   SE
4  2017-02-26 2017-02-26         20.929932           1   SE
5  2017-08-21 2017-08-21        -88.739990           0   SE
6  2018-02-15 2018-02-15        678.080078           1   SE
7  2018-07-13 2018-07-13          3.020020           1   SE
8  2018-08-11 2018-08-11        109.939941           1   SE
9  2019-01-06 2019-01-06        240.113525           1   SE
10 2019-07-02 2019-07-02        212.994141           1   SE
11 2019-12-26 2019-12-26        -35.832520           0   SE
12 2020-06-21 2020-06-21        -27.296875           0   SE
13 2020-12-14 2020-12-14        102.152344           1   SE
14 2021-06-10 2021-06-10       -686.917969           0   SE
15 2021-12-04 2021-12-04      -4527.175781           0   SE
16 2022-04-30 2022-04-30       -890.984375           0   SE
17 2022-10-25 2022-10-25        750.892578           1   SE

In [39]:
le_df = final_df(lunar_eclipse_df,price.copy(deep=True))
le_df

START DATE   END DATE  Changes in Price  Inc or Dec TYPE
0  2014-10-08 2014-10-08         16.824005           1   LE
1  2015-04-04 2015-04-04         -0.593994           0   LE
2  2015-09-28 2015-09-28          6.306000           1   LE
3  2016-03-09 2016-03-09          0.965973           1   LE
4  2016-03-23 2016-03-23         -0.120026           0   LE
5  2016-09-16 2016-09-16         -0.272949           0   LE
6  2017-02-11 2017-02-11         15.552002           1   LE
7  2017-08-07 2017-08-07        166.159912           1   LE
8  2018-01-31 2018-01-31        112.899414           1   LE
9  2018-07-27 2018-07-27        214.609863           1   LE
10 2019-01-21 2019-01-21        -24.340332           0   LE
11 2019-07-16 2019-07-16      -1419.011719           0   LE
12 2020-01-10 2020-01-10        288.246582           1   LE
13 2020-06-05 2020-06-05       -134.682617           0   LE
14 2020-07-05 2020-07-05        -52.148438           0   LE
15 2020-11-30 2020-11-30       1447.513672           1   LE
16 2021-05-26 2021-05-26        901.574219           1   LE
17 2021-11-19 2021-11-19       1223.449219           1   LE
18 2022-05-16 2022-05-16      -1441.457031           0   LE
19 2022-11-08 2022-11-08      -2059.400391           0   LE

In [40]:
# concatenation of 4 data frames for rest of analyzing
le_df["TYPE"] = 0    # lunar eclipse
se_df["TYPE"] = 1    # solar eclipse
ms_df["TYPE"] = 2    # meteor shower
e_df["TYPE"] = 3     # equinox
last = pd.concat([le_df, se_df, ms_df, e_df])
last.reset_index(drop=True)
last

START DATE   END DATE  Changes in Price  Inc or Dec  TYPE
0  2014-10-08 2014-10-08         16.824005           1     0
1  2015-04-04 2015-04-04         -0.593994           0     0
2  2015-09-28 2015-09-28          6.306000           1     0
3  2016-03-09 2016-03-09          0.965973           1     0
4  2016-03-23 2016-03-23         -0.120026           0     0
..        ...        ...               ...         ...   ...
12 2020-09-22 2020-09-22         78.835938           1     3
13 2021-03-20 2021-03-20        -18.617188           0     3
14 2021-09-22 2021-09-22       2896.554688           1     3
15 2022-03-20 2022-03-20       -943.582031           0     3
16 2022-09-23 2022-09-23       -114.761719           0     3

[150 rows x 5 columns]

We created a single data frame that holds all the data and calculated the contingency table using the "Inc or Dec" and "TYPE" columns in this data frame.

In [41]:
crosstab = pd.crosstab(last['Inc or Dec'],last['TYPE'], margins=True)
crosstab

TYPE         0   1   2   3  All
Inc or Dec                     
0            9   9  39   8   65
1           11   9  56   9   85
All         20  18  95  17  150

Pearson's Chi-Square statistical hypothesis is used to test independence between categorical variables are shown as row and column variables in the contingency table. Independence means that knowing the value of the column variable does not change the probabilities of the row variable (and vice versa).

**Hypothesis :**

>H0(Null Hypothesis) : There is no relation between astronomical events and bitcoin price.

>H1(Alternative Hypothesis) : There is a relation between astronomical events and bitcoin price.


We used the chi2_contingency() function of scipy.stats module takes as input, the contingency table in 2d array format. It returns a tuple containing test statistics, the p-value, degrees of freedom and expected table in that order. The expected table is expected frequencies, based on the marginal sums of the table.

In [42]:
from scipy.stats import chi2_contingency
stat, p, dof, expected = chi2_contingency(crosstab.iloc[0:2,0:4])
print("Chi-square value: ",stat)
print("p-value: ",p)
print("Degree of Freedom: ",dof)
print("Expected Table:\n",expected)

Chi-square value:  0.6457419833853995
p-value:  0.8858854921230519
Degree of Freedom:  3
Expected Table:
 [[ 8.66666667  7.8        41.16666667  7.36666667]
 [11.33333333 10.2        53.83333333  9.63333333]]


We used the chi2.ppf() function of scipy.stats module takes as input, the confidence level and degree of freedom in that order. It returns the critical value.

In [43]:
from scipy.stats import chi2
critical_value= chi2.ppf(q = 0.95, # q= (1-.05) 95% confidence*
                      df = 3)      # df= degree of freedom
print("Critical value:",critical_value)

Critical value: 7.814727903251179


There are two methods for verifying the hypothesis:

1- Using p-value:

If the p-value is greater than the significance value(a=0.05), then H0 is accepted. Otherwise, H1 is accepted.

>0.8858854921230519 > 0.05

>p-value > a-value ⇒ H0 is accepted.

2- Using chi-square value:

>chi-square value(X^2)  = ∑ ( (observed-expected)^2 / expected )

If the critical value is greater or equal to the chi-square value, then H0 is accepted. Otherwise, H1 is accepted.

>7.814727903251179 > 0.6457419833853995

>critical value > chi-square value ⇒ H0 is accepted.

## Conclusion<a class="anchor" id="conc"></a>

In both methods, we found the same solution which is “𝐻0 is accepted “. As a result, we can say that there is no relation between astronomical events and bitcoin prices. Therefore, astronomical events such as eclipses, meteor showers, and equinoxes do not affect bitcoin prices.

## References<a class="anchor" id="references"></a>

[1]https://towardsdatascience.com/7-steps-to-a-successful-data-science-project-b452a9b57149

[2]https://plotly.com/python/candlestick-charts/

[3]http://www.seasky.org/astronomy/astronomy-calendar-2022.html#Top 

[4]https://analyticsindiamag.com/a-beginners-guide-to-chi-square-test-in-python-from-scratch/

[5]https://www.geeksforgeeks.org/python-pearsons-chi-square-test/

**Disclaimer!** <font color='grey'>This notebook was prepared by Feyza Özkan and Hatice Kübra Tekin as a term project for the *BBM467 - Data Intensive Applications* class. The notebook is available for educational purposes only. There is no guarantee on the correctness of the content provided as it is a student work.

If you think there is any copyright violation, please let us [know](https://forms.gle/BNNRB2kR8ZHVEREq8). 
</font>